Knowledge base on runtime benchmarking

jimmychu0807 opened the pull request:
There will be a companion PR on writing a tutorial of runtime benchmarking, but I will separate them in another PR. 

This is the first draft on the knowledge base. Please correct for any technical accuracy as well as any english typo / grammar mistake.

If you prefer to read in a webpage format, it is also [deployed in staging environment](https://devhub-maindocs.herokuapp.com/docs/en/knowledgebase/runtime/benchmarking). Username and password is provided in devhub Element channel.

Thanks @shawntabrizi @danforbes @NukeManDan

shawntabrizi commented on yarn.lock:
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
this file probably didnt mean to exist?

shawntabrizi commented on docs/tutorials/runtime-benchmarking/index.md:
```diff
@@ -0,0 +1,5 @@
+title: Introduction
+---
+
+# Benchmark `pallet-template`
```
??

jimmychu0807 replied:
Keeping this file as placeholder for now so no dead link and pass the CI test.

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+take so many extrinsic calls and compute for their results. We use weight to represent how much
+computations are needed for a runtime extrinsic.
+
+Some extrinsics has more computation and has a larger weight. If this extrinsic is included, then
```
```suggestion
Some extrinsics have more computation and therefore have a larger weight. If this extrinsic is included, then
```

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+the [Transaction Weight chapter](../learn-substrate/weight).
+
+In Substrate, **10^12 Weight = 1 Second**, and i.e 1,000 weight = 1 nanosecond, measured on a
+referenced hardware.
```
need reference to what is the _reference hardware_.

jimmychu0807 replied:
added to the footnote.

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+database storage) when running the extrinsic. As any computer scientist can tell, we are getting
+into a time complexity analysis of the function.
+
+This is where Substrate Runtime Benchmarking comes in. It has a set of toolchain to help determine
```
```suggestion
This is where Substrate Runtime Benchmarking comes in. It has a set of toolchains to help determine
```

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+
+## Features of Substrate Benchamrking Toolchain
+
+- You can configure multiple input variables to be used when benchmarking an extrinsic. For each
```
would be good to link to the `benchmarks!` macro do somewhere here, it as very good documentation.

jimmychu0807 replied:
added on L#85

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+  computation time changes with respect to the change of your specified variables.
+
+- It generates a `WeightInfo` structure in Rust that can either be further tweaked, or easily
+  integrated and called in your runtime.
```
```suggestion
  integrated into your runtime.
```

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+        // Individual benchmark cases here
+      }
+
+      impl_benchmark_test_suite!(Pallet, crate::tests::new_test_ext(),
+        crate::tests::Test);
```
```suggestion
      impl_benchmark_test_suite!(
        Pallet, 
        crate::tests::new_test_ext(),
        crate::tests::Test,
      );
```

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+    set_dummy_benchmark {
+      let b in 1 .. 1000;
+    }: {
+      set_dummy (RawOrigin::Root, b.into());
```
```suggestion
      set_dummy(RawOrigin::Root, b.into());
```

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+    accumulate_dummy {
+      let b in 1 .. 1000;
+      let caller = account("caller", 0, 0);
+    }: _ (RawOrigin::Signed(caller), b.into())
```
```suggestion
    }: _(RawOrigin::Signed(caller), b.into())
```

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
````diff
@@ -0,0 +1,547 @@
+    ```
+
+    At the last line in the `benchmarks!` macro, `impl_benchmark_test_suite!` takes three input: the
+    module name, the function that generates a test genesis storage `new_text_ext()`, and a test
````
```suggestion
    module name, the function that generates a test genesis storage (i.e. `new_text_ext()`), and a test
```

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+
+    At the last line in the `benchmarks!` macro, `impl_benchmark_test_suite!` takes three input: the
+    module name, the function that generates a test genesis storage `new_text_ext()`, and a test
+    runtime we built when writing our test cases, and expand them into testing code.
```
```suggestion
    runtime, and expand them into testing code. The test runtime can often be the same one used for testing. 
```

This is better, because some benchmarks might need special mock-ing.

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+    module name, the function that generates a test genesis storage `new_text_ext()`, and a test
+    runtime we built when writing our test cases, and expand them into testing code.
+
+    Ultimately benchmarking code is expanded as testing code, so it takes the code you write in
```
I find this to be a bit confusing. The whole benchmarking does indeed _not_ expand to testing code, whereas the opposite, it should be run with `--release`. The testing portion is executed in a debug release.

As noted above, I think you can rework this section maybe, just to point out that: 
1. hey, you need a mock runtime for testing your benchmarks as well
2. hey, you might be able to re-use the one from your pallet testing
3. [additional links etc]

jimmychu0807 replied:
I just finish reading the [benchmark! doc](https://substrate.dev/rustdocs/v3.0.0/frame_benchmarking/macro.benchmarks.html) and want to clarify what `impl_benchmark_test_suite!` does. 

- So running `substrate benchmark` actually doesn't run the code generated by `impl_benchmark_test_suite!`?
- It generates code to be run in `cargo test`, and run with the high and low values of  complexity parameters and see if it returns an `Ok(())`?

kianenigma replied:
>  So running substrate benchmark actually doesn't run the code generated by impl_benchmark_test_suite!?

No, that only runs the main benchmark code. The tests are run if you go into your pallet and run `cargo test --all-features`, or if you have done: 

```
#[cfg(any(test, feature = "runtime-benchmakr))]
mod benchmarking;
``` 

Then any `cargo test` will run the benchmark code as well. 

Some pallets do this, some don't afaik. I personally like this. The only downside is that if a benchmark test takes a lot of time, it will pollute the main test suite. But this is a rare case.

kianenigma replied:
> It generates code to be run in cargo test, and run with the high and low values of complexity parameters and see if it returns an Ok(())?

Correct. I think it generates just 3 steps per benchmark, with repeat=1, where the 3 steps are [low, mid, high] of the range.

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+    pallet-you-named = { version = ..., default-features = false, path = ... }
+
+    [features]
+    std = [
+      # -- snip --
+      "pallet-you-named/std"
+    ]
```
Technically unrelated, but I agree that comparing this with std makes the learning experience better.

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+    Tutorial](../../tutorials/add-a-pallet). Then our runtime also implements
+    `frame_benchmarking::Benchmark` trait under a feature flag, and implement the
+    `dispatch_benchmark` method, as [shown in the
+    file](https://github.com/paritytech/substrate/blob/master/bin/node/runtime/src/lib.rs#L1412).
```
```suggestion
    file](https://github.com/paritytech/substrate/blob/master/bin/node/runtime/src/lib.rs#L1412).
```

This link will become invalid any day as it is not bound to a commit, and only line number.

jimmychu0807 replied:
updated.

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
````diff
@@ -0,0 +1,547 @@
+    ```
+
+    There seems to be a lot of code, but we are mainly setting up an environment that all
+    benchmarking code will run, such as we pass in a set of storage keys that should not be counted
````
```suggestion
    benchmarking code will run. For example, we pass in a set of storage keys that should not be counted
```

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
````diff
@@ -0,0 +1,547 @@
+[`frame/example/src/lib.rs`](https://github.com/paritytech/substrate/blob/master/frame/example/src/lib.rs):
+
+```rust
+pub trait WeightInfo {
````
hmm wait -- WeightInfo is also generated in as the output.

jimmychu0807 replied:
I believe by default when running benchmark as is, it doesn't generate the trait, but just the implementation. It is for Substrate now because the CI/developer is using [frame-weight-template.hbs](https://github.com/paritytech/substrate/blob/v3.0.0/.maintain/frame-weight-template.hbs) as the template to generate the `weightInfo` implementation by passing additional `--template` cli argument. 

I am not covering WeightInfo template here and just introducing the default generated `WeightInfo` impl output. 

@shawntabrizi Just want to confirm if my understanding is correct.

kianenigma replied:
You are correct but my question is why not cover this? I personally have found it a bit of pain to manually maintain correct signatures in `Trait WeightInfo` and the fact that it can be autogenerated, IMO, is important.

kianenigma commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,547 @@
+  // -- snip --
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    #[pallet::weight(<T as pallet::Config>::WeightInfo::accumulate_dummy(0))]
```
```suggestion
    #[pallet::weight(T::WeightInfo::accumulate_dummy(0))]
```

kianenigma reviewed (commented):
mostly good so far. What I would like to request is one additional example where a dispatch's weight is actually correlated with one of the parameters. in the `set_dummy`, there is this random `b` and maybe people will not learn what it _should have been_.

jimmychu0807 commented:
thank you for the detail review @kianenigma 🙏

jimmychu0807 commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,567 @@
+
+For each extrinsic, we add an attribute macro `pallet::weight` and pass in the weight implementation
+function with the same name as the extrinsic and variables that affect the weight. We can pass
+extrinsic parameters as weight function inputs to estimated the final weight.
```
@kianenigma see if this addresses your last comment.

shawntabrizi commented:
@jimmychu0807 i went through and made changes where i thought it made sense.

Please do one more check and then feel free to merge.

jimmychu0807 commented on docs/knowledgebase/runtime/benchmarking.md:
```diff
@@ -0,0 +1,561 @@
+> template file and the benchmarking toolchain will fill in the exact numbers from the measured
+> result. This enables automating weight generation in your desired code format and integrates this
+> in your CI process.
+
```
@kianenigma added a paragraph here telling readers they can use `--template` to customize the weightinfo implementation.

jimmychu0807 commented:
I am merging this PR.

@shawntabrizi thanks for the update.
@kianenigma I made some updates based on your feedback. See if we altogether have addressed your concerns. If not, we can address it in another PR.
